lib/posix/pthread_mutex.c

```diff
@@ -9,16 +9,214 @@
 #include <wait_q.h>
 #include <posix/pthread.h>
 
-int pthread_mutex_trylock(pthread_mutex_t *m)
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_mutex_sem, sizeof(struct k_sem),
+		  sizeof(struct k_sem), 32, 4);
+
+/*
+ *  Default mutex attrs.
+ */
+static const pthread_mutexattr_t def_attr = {
+	.type = PTHREAD_MUTEX_DEFAULT,
+	.protocol = PTHREAD_PRIO_NONE,
+};
+
+static int mutex_recursive_take(pthread_mutex_t *m, int timeout)
+{
+	if (m->lock_count == 0) {
+		m->owner = pthread_self();
+	}
+
+	if (m->owner == pthread_self()) {
+		m->lock_count++;
+		if (m->lock_count > MUTEX_MAX_REC_LOCK) {
+			return EAGAIN;
+		}
+		return k_sem_take(m->sem, timeout);
```

Comment:
k_sem_take() returns a negative error number or 0 on success. You have to convert this to a positive error code or 0: `return -k_sem_take(m->sem, timeout)`

Reply:
Also, same comment about this function being called with IRQ locked and timeout being possibly K_FOREVER